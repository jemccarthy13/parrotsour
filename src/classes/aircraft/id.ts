export enum IDMatrix {
  HOSTILE = "red",
  SUSPECT = "orange",
  ASSUME_FRIEND = "green",
  NEUTRAL = "purple",
  FRIEND = "blue",
}

export function getMostRestrictiveID(ids: IDMatrix[]): IDMatrix {
  function idToNum(id: IDMatrix) {
    const num = 4
    switch (id) {
      case IDMatrix.SUSPECT:
        return 3
      case IDMatrix.NEUTRAL:
        return 2
      case IDMatrix.ASSUME_FRIEND:
        return 1
      case IDMatrix.FRIEND:
        return 1
    }
    return num
  }
  let lowestID = IDMatrix.HOSTILE
  let lowestIDNum = 4
  for (let i = 0; i < ids.length; i++) {
    const idNum = idToNum(ids[i])
    if (idNum < lowestIDNum) {
      lowestIDNum = idNum
      lowestID = ids[i]
    }
  }
  return lowestID
}
