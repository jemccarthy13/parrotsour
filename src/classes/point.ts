import { BlueInThe } from "../canvas/canvastypes"
import { BRAA } from "./braa"
import { PIXELS_TO_NM, toDegrees } from "../utils/psmath"

/**
 * An x,y Cartesian point that can calculate bearing and
 * range to another point.
 */
export class Point {
  x: number
  y: number

  static DEFAULT = new Point(0, 0)

  /**
   * Construct a new x,y point
   */
  constructor(x: number, y: number) {
    this.x = x
    this.y = y
  }

  /**
   * Get a bearing and range between this point and another point
   *
   * @param ptTo the "other" or "to" point
   * @returns the bearing and range between this (ptFrom) and ptTo
   */
  getBR(toPoint: Point): BRAA {
    const deltaX = this.x - toPoint.x
    const deltaY = this.y - toPoint.y

    // distance formula for range
    const rng = Math.floor(
      Math.sqrt(deltaX * deltaX + deltaY * deltaY) / PIXELS_TO_NM
    )

    // convert cartesian angle to heading
    let brg = Math.round(
      270 + toDegrees(Math.atan2(this.y - toPoint.y, this.x - toPoint.x))
    )
    if (brg > 360) {
      brg = brg - 360
    }

    return new BRAA(brg, rng)
  }

  /**
   * Get distance between points (straight along axis, i.e. not slant range)
   */
  public straightDistNM(toPoint: Point, orientation: BlueInThe): number {
    return Math.floor(
      Math.abs(
        orientation === BlueInThe.NORTH || orientation === BlueInThe.SOUTH
          ? this.y - toPoint.y
          : this.x - toPoint.x
      ) / PIXELS_TO_NM
    )
  }
}
