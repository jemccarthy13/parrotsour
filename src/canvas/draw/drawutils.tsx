/**
 * This file contains common low-level drawing utilities used
 * to build groups and pictures.
 */

// Interfaces
import { Point } from "../../classes/point"

/**
 * 'Clamp' the location to the confines of the drawing context
 * @param ctx The context to constrict it to
 * @param pos the current position
 */
export function clampInContext(
  ctx: CanvasRenderingContext2D,
  x: number | Point,
  y?: number
): Point {
  let retPoint = Point.DEFAULT
  if (x instanceof Point) {
    retPoint = new Point(
      Math.min(Math.max(x.x, 0), ctx.canvas.width),
      Math.min(Math.max(x.y, 0), ctx.canvas.height)
    )
  } else {
    if (y === undefined) y = 0
    retPoint = new Point(
      Math.min(Math.max(x, 0), ctx.canvas.width),
      Math.min(Math.max(y, 0), ctx.canvas.height)
    )
  }
  return retPoint
}
