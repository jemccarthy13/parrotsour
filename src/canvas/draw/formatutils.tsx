/**
 * This file contains utilities for group and answer formatting
 */
import { AircraftGroup } from "../../classes/groups/group"
import { toRadians } from "../../utils/psmath"
import { ArrowDataTrail } from "../../classes/aircraft/datatrail/arrowdatatrail"

/**
 * Convert an altitude to a 3-digit flight level
 * @param alt int altitude to change to padded string
 */
export function formatAlt(alt: number): string {
  return (alt * 10).toString().substring(0, 3).padStart(3, "0")
}

/**
 * Check if picture is opening or closing
 * @param fg First group of picture
 * @param sg Second group of picture
 */
export function getOpenCloseAzimuth(
  fg: AircraftGroup,
  sg: AircraftGroup
): string {
  let retVal = ""

  const fgStartPos = fg.getStartPos()
  const sgStartPos = sg.getStartPos()

  const angle1to2 = fgStartPos.getBR(sgStartPos).bearingNum

  const fgHeading = fg.getHeading()
  const sgHeading = sg.getHeading()

  const speed1to2 =
    ArrowDataTrail.LEN_TRAIL * Math.cos(toRadians(angle1to2 - fgHeading))

  const speed2to1 =
    ArrowDataTrail.LEN_TRAIL * Math.cos(toRadians(angle1to2 - sgHeading))

  const acceptableRate = Math.cos(toRadians(5))

  let closing = speed1to2 > 0 && speed2to1 < 0
  let opening = speed1to2 < 0 && speed2to1 > 0

  closing =
    closing ||
    (Math.abs(speed1to2) < acceptableRate && speed2to1 < 0) ||
    (Math.abs(speed2to1) < acceptableRate && speed1to2 > 0)
  opening =
    opening ||
    (Math.abs(speed1to2) < acceptableRate && speed2to1 > 0) ||
    (Math.abs(speed2to1) < acceptableRate && speed1to2 < 0)

  const parallel =
    Math.abs(speed1to2) < acceptableRate && Math.abs(speed2to1) < acceptableRate

  if (!parallel) {
    if (opening) {
      retVal = " OPENING "
    }
    if (closing) {
      retVal = " CLOSING "
    }
  }

  return retVal
}
