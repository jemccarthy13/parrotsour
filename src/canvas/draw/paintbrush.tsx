import { Braaseye } from "../../classes/braaseye"
import { AircraftGroup } from "../../classes/groups/group"
import { Point } from "../../classes/point"
import { PIXELS_TO_NM, randomNumber } from "../../utils/psmath"
import { PictureCanvasProps, PictureCanvasState } from "../canvastypes"
import { clampInContext } from "./drawutils"
import { formatAlt } from "./formatutils"

export class PaintBrush {
  private static ctx: CanvasRenderingContext2D

  constructor(ctx: CanvasRenderingContext2D) {
    PaintBrush.ctx = ctx
  }

  public static use(ctx: CanvasRenderingContext2D | undefined | null): void {
    if (ctx) PaintBrush.ctx = ctx
  }

  public static getContext(): CanvasRenderingContext2D {
    return PaintBrush.ctx
  }

  public static clearCanvas(): void {
    const ctx = PaintBrush.ctx
    ctx.clearRect(0, 0, ctx.canvas.width, ctx.canvas.height)
    ctx.fillStyle = "white"
    ctx.fillRect(0, 0, ctx.canvas.width, ctx.canvas.height)
  }

  /**
   * Draw a line on the drawing context, given the properties
   * @param startX start x of the line
   * @param startY start y of the line
   * @param endX end x of the line
   * @param endY end y of the line
   * @param color (optional) color of the line, defaults to black
   */
  public static drawLine(
    startX: number,
    startY: number,
    endX: number,
    endY: number,
    color = "black",
    forcedCtx?: CanvasRenderingContext2D
  ): void {
    let context = PaintBrush.ctx
    if (forcedCtx) context = forcedCtx
    context.lineWidth = 1
    context.strokeStyle = color
    context.beginPath()
    context.moveTo(startX, startY)
    context.lineTo(endX, endY)
    context.stroke()
    context.stroke()
  }

  /**
   * Draw text to the drawing context at the given position
   * @param text Text to draw
   * @param x X Position to draw at
   * @param y Y Position to draw at
   * @param size (optional) Size of the text to draw, defaults 12
   * @param color (optional) Color of the text, defaults to black
   */
  public static drawText(
    text: string,
    x: number,
    y: number,
    size = 12,
    color = "black"
  ): void {
    const ctx = PaintBrush.ctx
    ctx.lineWidth = 1
    ctx.fillStyle = color
    ctx.font = size + "px Arial"
    const pos = clampInContext(ctx, x, y)

    ctx.fillText(text, pos.x, pos.y)
    ctx.fillText(text, pos.x, pos.y)
  }

  /**
   * Draw a measurement (distance with a length number)
   * @param {Point} start position to start for line
   * @param {Point} end position to end for line
   * @param {number} distance how long the line should be
   * @param {boolean} showMeasurements Display the measurements
   */
  public static drawMeasurement(
    start: Point,
    end: Point,
    distance: number,
    showMeasurements: boolean
  ): void {
    if (showMeasurements) {
      PaintBrush.drawLine(start.x, start.y, end.x, end.y)
      PaintBrush.drawText(
        Math.floor(distance).toString(),
        (start.x + end.x) / 2,
        (start.y + end.y) / 2 - 3
      )
    }
  }

  /**
   * Draw altitudes next to a group with optional offset
   *
   * @param grpPos group's current position
   * @param alts altitudes to draw
   * @param offX (optional) x-axis offset from group
   * @param offY (optional) y-axis offset from group
   */
  public static drawAltitudes(
    grpPos: Point,
    alts: number[],
    offX?: number,
    offY?: number
  ): void {
    const offsetX = offX || 0
    const offsetY = offY || 0
    const formattedAlts: string[] = alts.map((a: number) => {
      return formatAlt(a)
    })
    PaintBrush.drawText(
      formattedAlts.join(","),
      grpPos.x + 25 + offsetX,
      grpPos.y - 11 + offsetY,
      11,
      "#ff8c00"
    )
  }

  public static drawBullseye(bull?: Point, color?: string): Point {
    const context = PaintBrush.ctx
    color = color || "black"
    context.lineWidth = 1
    context.fillStyle = color
    context.strokeStyle = color

    const centerPointX = bull
      ? bull.x
      : randomNumber(context.canvas.width * 0.33, context.canvas.width * 0.66)
    const centerPointY = bull
      ? bull.y
      : randomNumber(context.canvas.height * 0.33, context.canvas.height * 0.66)

    context.beginPath()
    context.arc(
      centerPointX,
      centerPointY,
      PIXELS_TO_NM / 2,
      0,
      2 * Math.PI,
      true
    )
    context.stroke()
    context.fill()

    context.moveTo(centerPointX, centerPointY + PIXELS_TO_NM * 2)
    context.lineTo(centerPointX, centerPointY - PIXELS_TO_NM * 2)
    context.stroke()

    context.moveTo(centerPointX + PIXELS_TO_NM * 2, centerPointY)
    context.lineTo(centerPointX - PIXELS_TO_NM * 2, centerPointY)
    context.stroke()

    return new Point(centerPointX, centerPointY)
  }

  public static drawFullInfo(
    state: PictureCanvasState,
    props: PictureCanvasProps,
    groups: AircraftGroup[]
  ): void {
    for (let y = 0; y < groups.length; y++) {
      const grpPos = groups[y].getCenterOfMass(props.dataStyle)
      if (props.showMeasurements) {
        new Braaseye(
          grpPos,
          state.blueAir.getCenterOfMass(props.dataStyle),
          state.bullseye
        ).draw(true, props.braaFirst)
      }
      PaintBrush.drawAltitudes(grpPos, groups[y].getAltitudes())
    }
  }
}
