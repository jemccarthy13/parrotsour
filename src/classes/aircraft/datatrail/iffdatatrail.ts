import { PaintBrush } from "../../../canvas/draw/paintbrush"
import { headingToRadians, PIXELS_TO_NM } from "../../../utils/psmath"
import { Point } from "../../point"
import { IDMatrix } from "../id"
import { DataTrail } from "./datatrail"

export class IFFDataTrail extends DataTrail {
  private iffPoints: Point[] = []
  private LEN_TRAIL: number = 7 * PIXELS_TO_NM

  constructor(startPos: Point, heading: number) {
    super(startPos)

    const starty = startPos.y
    const startx = startPos.x
    const vector = headingToRadians(heading)
    const endy = starty + this.LEN_TRAIL * -Math.sin(vector.radians)
    const endx = startx + this.LEN_TRAIL * Math.cos(vector.radians)

    const offsetX = (endx - startx) / PIXELS_TO_NM
    const offsetY = (endy - starty) / PIXELS_TO_NM

    let xPos = startx
    let yPos = starty
    const iPts: Point[] = []
    // draw the radar trail
    if (!this.iffPoints || this.iffPoints.length === 0) {
      for (let mult = 0; mult < 5; mult++) {
        xPos = startx + offsetX * mult + offsetX * 0.5
        yPos = starty + offsetY * mult + offsetY * 0.5
        iPts.push(new Point(xPos, yPos))
      }
      this.iffPoints = iPts
    }
  }

  getCenterOfMass(): Point {
    return this._getOnePlotAhead()
  }

  public move(heading: number): void {
    // this logic should be in move()
    // take the first point out of the radar trail
    const newIFFPts = this.iffPoints.slice(1)

    const startPos = this.getStartPos()
    const startx = startPos.x
    const starty = startPos.y
    const vector = headingToRadians(heading)

    const endy = starty + this.LEN_TRAIL * -Math.sin(vector.radians)
    const endx = startx + this.LEN_TRAIL * Math.cos(vector.radians)
    const offsetX = (endx - startx) / PIXELS_TO_NM
    const offsetY = (endy - starty) / PIXELS_TO_NM

    // add a bit of jitter with randomness
    const jit = 1

    const xPos = startx + offsetX * 6 + jit * Math.random() + Math.random()
    const yPos = starty + offsetY * 6 + jit * Math.random() + Math.random()
    newIFFPts.push(new Point(xPos, yPos))
    this.iffPoints = newIFFPts
  }

  draw(heading: number, id: IDMatrix): void {
    const ctx = PaintBrush.getContext()
    if (id !== IDMatrix.SUSPECT && id !== IDMatrix.HOSTILE) {
      ctx.strokeStyle = "blue"
      ctx.lineWidth = 1
      for (let l = 0; l < this.iffPoints.length; l++) {
        const xPos = this.iffPoints[l].x
        const yPos = this.iffPoints[l].y
        ctx.beginPath()
        ctx.moveTo(xPos, yPos)
        ctx.lineTo(xPos - 3, yPos)
        ctx.lineTo(xPos - 3, yPos - 3)
        ctx.lineTo(xPos, yPos - 3)
        ctx.lineTo(xPos, yPos)
        ctx.stroke()
      }
    }
  }

  _getOnePlotAhead(): Point {
    const cPt = this.iffPoints[this.iffPoints.length - 1]
    const pPt = this.iffPoints[this.iffPoints.length - 2]

    const deltX = cPt.x - pPt.x
    const deltY = cPt.y - pPt.y

    return new Point(cPt.x + deltX, cPt.y + deltY)
  }
}
