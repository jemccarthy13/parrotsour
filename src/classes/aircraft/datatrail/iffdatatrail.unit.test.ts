import { PaintBrush } from "../../../canvas/draw/paintbrush"
import { Point } from "../../point"
import { IDMatrix } from "../id"
import { IFFDataTrail } from "./iffdatatrail"

// eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-unused-vars
const canvasSerializer = require("jest-canvas-snapshot-serializer")
expect.addSnapshotSerializer(canvasSerializer)

const canvas = document.createElement("canvas")
// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
PaintBrush.use(canvas.getContext("2d")!)

describe("IFFDataTrail", () => {
  it("draws_correctly", () => {
    const datatrail = new IFFDataTrail(new Point(50, 50), 90)
    datatrail.draw(90, IDMatrix.FRIEND)
    expect(canvas).toMatchSnapshot()
  })

  it("doesnt_draw_hos_sus", () => {
    const datatrail = new IFFDataTrail(new Point(50, 50), 90)
    datatrail.draw(90, IDMatrix.HOSTILE)
    datatrail.draw(90, IDMatrix.SUSPECT)
    expect(canvas).toMatchSnapshot()
  })
})
