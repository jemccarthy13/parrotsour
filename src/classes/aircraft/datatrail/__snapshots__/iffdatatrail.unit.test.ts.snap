// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IFFDataTrail doesnt_draw_hos_sus 1`] = `
<canvas
  data-snapshot-image="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAACWCAYAAABkW7XSAAAABmJLR0QA/wD/AP+gvaeTAAABAElEQVR4nO3XSwqAMAwE0Fa8mCdTT163Kn4QUgLy3mrCbGabUgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA4E3NHhCnLce7Lnkd0MOQPSDQfJMzOoAnrV3njA7oYcweEOv8pmV2QLQ/vYTrLk/JHQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAADAFxtHvi4xtxq3hQAAAABJRU5ErkJggg=="
/>
`;

exports[`IFFDataTrail draws_correctly 1`] = `
<canvas
  data-snapshot-image="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAACWCAYAAABkW7XSAAAABmJLR0QA/wD/AP+gvaeTAAABAElEQVR4nO3XSwqAMAwE0Fa8mCdTT163Kn4QUgLy3mrCbGabUgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA4E3NHhCnLce7Lnkd0MOQPSDQfJMzOoAnrV3njA7oYcweEOv8pmV2QLQ/vYTrLk/JHQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAADAFxtHvi4xtxq3hQAAAABJRU5ErkJggg=="
/>
`;
